extension/src/topn.rs

```diff
@@ -0,0 +1,325 @@
+    stype = internal,
+    finalfunc = toolkit_experimental.topn_final,
+    combinefunc = toolkit_experimental.topn_combine,
+    serialfunc = toolkit_experimental.topn_serialize,
+    deserialfunc = toolkit_experimental.topn_deserialize,
+    parallel = safe
+);
+"#);
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn topn_compound_trans(
+    state: Option<Internal<InternalTopN>>,
+    value: Option<toolkit_experimental::TopN<'static>>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<InternalTopN>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state, value) {
+                (a, None) => a,
+                (None, Some(a)) => Some(a.to_internal_topn().into()),
+                (Some(a), Some(b)) =>
+                    Some(InternalTopN::combine(&a, &b.to_internal_topn()).into()),
+            }
+        })
+    }
+}
+
+extension_sql!(r#"
+CREATE AGGREGATE toolkit_experimental.rollup(
+    toolkit_experimental.topn
+) (
+    sfunc = toolkit_experimental.topn_compound_trans,
+    stype = internal,
+    finalfunc = toolkit_experimental.topn_final,
+    combinefunc = toolkit_experimental.topn_combine,
+    serialfunc = toolkit_experimental.topn_serialize,
+    deserialfunc = toolkit_experimental.topn_deserialize,
+    parallel = safe
+);
+"#);
+
+//---- Available PG operations on the topn structure
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn num_vals(
+    agg: toolkit_experimental::TopN<'static>,
+) -> int {
+    agg.total_inputs as _
+}
+
+#[pg_extern(immutable, parallel_safe, name="topn", schema = "toolkit_experimental")]
+pub fn topn_iter (
+    n: i32,
+    agg: toolkit_experimental::TopN<'static>,
+) -> impl std::iter::Iterator<Item = (name!(value,i64),name!(min_freq,f64),name!(max_freq,f64))> + '_ {
+    assert!(n <= agg.num_values as _);
+    let mut count = 0;
+    let total = agg.total_inputs as f64;
+    Box::new(multizip((agg.values.iter(), agg.counts.iter(), agg.overcounts.iter())).filter(move |_| {count = count + 1; count <= n}).map(move |x| {let (val, count, over) = x; (*val, (count-over) as f64 / total, *count as f64 / total)}))
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn guaranteed_topn(
+    n: i32,
+    agg: toolkit_experimental::TopN<'static>,
+) -> bool {
+    if n >= agg.num_values as _ {
+        return false;
+    }
+
+    let bound = agg.counts[n as usize];
+    for i in 0..n as usize {
+        if agg.counts[i] - agg.overcounts[i] < bound {
+            return false;
+        }
+    }
+
+    true
+}
+
+#[pg_extern(immutable, parallel_safe, schema = "toolkit_experimental")]
+pub fn max_ordered_n(
+    agg: toolkit_experimental::TopN<'static>,
+) -> int {
+    for i in 1..agg.num_values as usize {
+        if agg.counts[i] > agg.counts[i-1] - agg.overcounts[i-1] {
+            return (i - 1) as _;
+        }
+    }
+
+    (agg.num_values - 1) as _
+}
```

Comment:
Note for the future: we probably want a `top_guaranteed()` which outputs elements until this loop condition is false.

Reply:
Agreed.